network/wsPeer.go

```diff
@@ -393,6 +429,33 @@ func dedupSafeTag(t protocol.Tag) bool {
 	return t == protocol.AgreementVoteTag || t == protocol.TxnTag
 }
 
+// MaxDecompressedMessageSize defines a maximum decompressed data size
+// to prevent zip bombs
+const MaxDecompressedMessageSize = 20 * 1024 * 1024 // some large enough value
+
+func decompressMsg(data []byte) ([]byte, error) {
+	r := zstd.NewReader(bytes.NewReader(data))
+	b := make([]byte, 0, 1024)
+	for {
+		if len(b) == cap(b) {
+			b = append(b, 0)[:len(b)]
+		}
+		n, err := r.Read(b[len(b):cap(b)])
+		b = b[:len(b)+n]
+		if err != nil {
+			r.Close()
```

Comment:
Maybe cleaner to turn this `r.Close()` and the one below into a `defer r.Close()` after the call to `zstd.NewReader()`?

Reply:
Right, originally it was inside the `readLoop` function and defer would not work. Fixed.